### PR TITLE
feat: Auth codegen

### PR DESCRIPTION
**Pull request by Sanayshah2**
#### Description of changes
Added OAuth flows, Oauth scopes, Unauthenticated logins, Custom Attributes, Group precedence, read/write permissions for attributes, User pool name, oidc/saml external providers codegen

Some features are natively supported; for those not supported natively, Property Overrides are used to mimic the same behavior.

#### Description of how you validated changes
Unite tests for various scenarios and edge cases. Also, manually verified against a deployed application and checking for generated code having relevant property overrides.
#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#tests)
- [x] Relevant documentation is changed or added (and PR referenced)
- [x] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies
- [x] [Pull request labels](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#labels) are added

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -68,7 +68,7 @@ module.exports = {
       'warn',
       {
         lang: 'en_US',
-        skipWords: dictionary,
+        skipWords: dictionary.concat(['saml']), // Add 'saml' to the skipWords array
```
Comment (sobolk) [inline .eslintrc.js:71]: Can you add this to https://github.com/aws-amplify/amplify-cli/blob/migrations/.eslint-dictionary.json instead?

File: packages/amplify-gen2-codegen/API.md [hunk 2]
```diff
@@ -140,8 +186,10 @@ export type StorageTriggerEvent = 'onDelete' | 'onUpload';
 // @public (undocumented)
 export type UserPoolMfaConfig = 'OFF' | 'ON' | 'OPTIONAL';
 
-// @public (undocumented)
-export type UserPoolOverrides = Partial<Record<PasswordPolicyPath, string | number | boolean>>;
+// Warnings were encountered during analysis:
+//
+// src/auth/source_builder.ts:73:3 - (ae-forgotten-export) The symbol "MetadataOptions" needs to be exported by the entry point index.d.ts
+// src/auth/source_builder.ts:86:3 - (ae-forgotten-export) The symbol "OidcEndPoints" needs to be exported by the entry point index.d.ts
```
Comment (sobolk) [inline packages/amplify-gen2-codegen/API.md:192]: this should be addressed.

File: packages/amplify-gen2-codegen/src/auth/source_builder.ts [hunk 3]
```diff
@@ -34,27 +49,57 @@ export type UserPoolMfaConfig = 'OFF' | 'ON' | 'OPTIONAL';
 
 export type PasswordPolicyPath = `Policies.PasswordPolicy.${keyof PasswordPolicyType}`;
 
-export type UserPoolOverrides = Partial<Record<PasswordPolicyPath, string | number | boolean>>;
+export type PolicyOverrides = Partial<Record<PasswordPolicyPath | string, string | boolean | number | string[]>>;
+
+// export type UserPoolOverrides = Partial<Record<PasswordPolicyPath | string, string | number | boolean | string[]>>;
```
Comment (sobolk) [inline packages/amplify-gen2-codegen/src/auth/source_builder.ts:54]: please remove if not needed.

**Review by sobolk — COMMENTED** [review]
looks good.

File: packages/amplify-gen2-codegen/API.md (hunk 2 above)
Comment (Sanayshah2) [reply packages/amplify-gen2-codegen/API.md]: Resolved - [60d2ac9](https://github.com/aws-amplify/amplify-cli/pull/13913/commits/60d2ac94878b76ac7627ea01c51058cbc42324ef)

File: packages/amplify-gen2-codegen/src/auth/source_builder.ts (hunk 3 above)
Comment (Sanayshah2) [reply packages/amplify-gen2-codegen/src/auth/source_builder.ts]: Done.

**Review by sobolk — APPROVED** [review]
